virt: add very initial virtualization support

lorc opened the pull request:
This patch enables vary basic virtualization support in OP-TEE.

Added new configuration variable CFG_VIRTUALIZATION. If it is
enabled, OP-TEE will be built with virtualization support. It will
rely on hypervisor, which shold inform OP-TEE about a new guest
using OPTEE_SMC_VM_CREATED call.

Only one client is allowed right now and there is no way to
unregister it.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         documentation/github.md.

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" part:
         https://github.com/OP-TEE/optee_os/blob/master/Notice.md#contributions.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jenswi-linaro commented on core/arch/arm/kernel/virtualization.c:
```diff
@@ -0,0 +1,32 @@
+
+static uint16_t current_client_id = 0;
+
+uint32_t client_created(uint16_t client_id)
```
It's guests we're dealing with here, I'd prefer `virt_guest_created()` and so on for the other functions

lorc replied:
Ah. Okay. I used "client" because SMCCC uses this term for virtual OSes and also it is used in `struct thread_ctx`.

I'll rework my PR and also I'll renamed field `hyp_clnt_id` in `struct thread_ctx` for uniformity.

jenswi-linaro replied:
I don't see a problem with letting the parameter have the name `client_id` as that's the name of the field. It's just that the function isn't keeping track of clients (using that terminology here would be very confusing).

jenswi-linaro commented on core/arch/arm/kernel/virtualization.c:
```diff
@@ -0,0 +1,32 @@
+
+uint32_t client_created(uint16_t client_id)
+{
+	if (current_client_id != 0)
```
This variable needs to be protected by a sync object.

lorc commented:
Addressed comments.

etienne-lms commented on core/arch/arm/kernel/virtualization.c:
```diff
@@ -0,0 +1,48 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+
```
missing copyright

etienne-lms commented on core/arch/arm/kernel/virtualization.c:
```diff
@@ -0,0 +1,48 @@
+	cpu_spin_lock(&client_id_lock);
+
+	if (current_client_id == client_id)
+		panic("Hypervisor tries to remove client");
```
why panicking here?
hypervisor should panic but not the tee.

etienne-lms commented on core/include/kernel/virtualization.h:
```diff
@@ -0,0 +1,13 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+
```
copyright

lorc commented:
Added fix

etienne-lms commented on core/arch/arm/kernel/virtualization.c:
```diff
@@ -0,0 +1,51 @@
+
+#define INVALID_CLIENT_ID		0xFFFF
+
+static uint16_t current_client_id = 0;
```
minor: `0` is a specific ID for the hypervisor, right?
Maybe init `current_client_id` to `INVALID_CLIENT_ID`. Or discard need for `INVALID_CLIENT_ID`.

etienne-lms commented on core/arch/arm/kernel/virtualization.c:
```diff
@@ -0,0 +1,51 @@
+
+bool check_virt_guest(uint16_t client_id)
+{
+	return client_id == 0 || client_id == current_client_id;
```
minor: could use `!client_id`

etienne-lms commented on core/arch/arm/tee/entry_fast.c:
```diff
@@ -155,6 +156,33 @@ static void tee_entry_boot_secondary(struct thread_smc_args *args)
+	uint16_t client_id = args->a1;
+
+	/* Only hypervisor can issue this request */
+	if (args->a7 != 0) {
```
minor: could use `if (args->a7)`, here and below.

lorc commented:
Hi @etienne-lms,

I addressed both your comments. I added special define `HYP_CLIENT_ID` to get rid of magic numbers.

As for `if (value) / if (value != 0)` - I prefer clarity over code length. So, if `value` holds some boolean type (even if it is declared as `unsigned  int`), I will write `if (value) / if (!value)`.  Example:

```
unsigned long addr_page_aligned = pa & SMALL_PAGE_MASK;
if (addr_page_aligned) {
...
}

```
But if `value` holds some integral value, like count of someting, then I'll prefer to compare against zero.  Example:

```
unsigned int num_bytes_left_on_page = size_left & SMALL_PAGE_MASK;
if (num_bytes_left_on_page != 0) {
...
}
```

Obviously, code in both examples will be compiled into exactly the same processor instructions. But first code sample emphasises that we are dealing with some boolean value, where in the second example - with count. 
This is subtle difference. But think of the following: you _can_ change `if (bytes_left_on_page != 0)` to` if (bytes_left_on_page < SOME_THRESHOLD)`. But there are nothing you can do with `if (addr_page_aligned)`. It is either true or false.

jenswi-linaro commented:
Regardless of style, both ways are usually just as clear when you're used to them. Sometimes a more compact representation makes it easier to get an overview of the expression. I think we tend to prefer the more compact representation (especially if it doesn't involve negative logic), but both are usually OK. This is as far as I know in line with the Linux kernel style too.

lorc commented:
Closing as a stale.
